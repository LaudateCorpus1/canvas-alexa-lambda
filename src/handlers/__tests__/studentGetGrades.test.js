/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that they will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
const axios = require("axios");
const MockAdapter = require("axios-mock-adapter");
const createVirtualAlexa = require("../../test_utils/utils.js").createVirtualAlexa;

let mock;
let alexa;

beforeAll(() => {
  alexa = createVirtualAlexa();
  mock = new MockAdapter(axios);
});

afterEach(() => {
  mock.reset();
});

afterAll(() => {
  mock.restore();
});

it("tells you if you are in no courses", async () => {
  const url = "/courses?enrollment_state=active&include[]=total_scores&enrollment_type=student";
  mock.onGet(url).reply(200, []);
  const result = await alexa.utter("What are my grades");
  expect(result).toMatchSnapshot();
});

it("tells you grades for multiple courses", async () => {
  const url = "/courses?enrollment_state=active&include[]=total_scores&enrollment_type=student";
  mock.onGet(url).reply(200, [
    {
      name: "test course 1",
      enrollments: [{ computed_current_grade: "A-" }]
    },
    {
      name: "test course 2",
      enrollments: [{ computed_current_score: "91%" }]
    },
    {
      name: "test course 3",
      enrollments: []
    }
  ]);
  const result = await alexa.utter("What are my grades");
  expect(result).toMatchSnapshot();
});

it("tells you your grades for a specific class using computed grades", async () => {
  const url = "/courses?enrollment_state=active&include[]=total_scores&enrollment_type=student";
  mock.onGet(url).reply(200, [
    {
      name: "Calculus",
      enrollments: [{ computed_current_grade: "A-" }]
    },
    {
      name: "Writing",
      enrollments: [{ computed_current_score: "91%" }]
    }
  ]);
  const result = await alexa.utter("What is my Calculus grade");
  expect(result).toMatchSnapshot();
});

it("tells you your grades for a specific class using computed grades", async () => {
  const url = "/courses?enrollment_state=active&include[]=total_scores&enrollment_type=student";
  mock.onGet(url).reply(200, [
    {
      name: "Calculus",
      enrollments: [{ computed_current_score: "A-" }]
    },
    {
      name: "Writing",
      enrollments: [{ computed_current_score: "91%" }]
    }
  ]);
  const result = await alexa.utter("What is my Writing grade");
  expect(result).toMatchSnapshot();
});

it("tells you no grades posted yet for a specific class", async () => {
  const url = "/courses?enrollment_state=active&include[]=total_scores&enrollment_type=student";
  mock.onGet(url).reply(200, [
    {
      name: "Calculus",
      enrollments: []
    },
    {
      name: "Writing",
      enrollments: [{ computed_current_score: "91%" }]
    }
  ]);
  const result = await alexa.utter("What is my Calculus grade");
  expect(result).toMatchSnapshot();
});

it("does not tels you your grades for a specific class you are not in", async () => {
  const url = "/courses?enrollment_state=active&include[]=total_scores&enrollment_type=student";
  mock.onGet(url).reply(200, [
    {
      name: "Calculus",
      enrollments: [{ computed_current_grade: "A-" }]
    },
    {
      name: "Writing",
      enrollments: [{ computed_current_score: "91%" }]
    }
  ]);
  const result = await alexa.utter("What is my Banana grade");
  expect(result).toMatchSnapshot();
});
