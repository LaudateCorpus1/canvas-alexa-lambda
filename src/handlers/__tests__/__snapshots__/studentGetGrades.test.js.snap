// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not tels you your grades for a specific class you are not in 1`] = `
SkillResponse {
  "response": Object {
    "outputSpeech": Object {
      "ssml": "<speak> No matching course found. Anything else? </speak>",
      "type": "SSML",
    },
    "reprompt": Object {
      "outputSpeech": Object {
        "ssml": "<speak> You can ask to list grades, or check if you have any missing assignments. </speak>",
        "type": "SSML",
      },
    },
    "shouldEndSession": false,
  },
  "sessionAttributes": Object {},
  "userAgent": "ask-nodejs/1.0.25 Node/v8.11.4",
  "version": "1.0",
}
`;

exports[`tells you grades for multiple courses 1`] = `
SkillResponse {
  "response": Object {
    "outputSpeech": Object {
      "ssml": "<speak> Here are your grades:
for test course 1: Your grade is A- .
for test course 2: Your grade is 91% .
for test course 3: Your grade is not posted yet .. Anything else? </speak>",
      "type": "SSML",
    },
    "reprompt": Object {
      "outputSpeech": Object {
        "ssml": "<speak> You can ask to list grades, or check if you have any missing assignments. </speak>",
        "type": "SSML",
      },
    },
    "shouldEndSession": false,
  },
  "sessionAttributes": Object {},
  "userAgent": "ask-nodejs/1.0.25 Node/v8.11.4",
  "version": "1.0",
}
`;

exports[`tells you if you are in no courses 1`] = `
SkillResponse {
  "response": Object {
    "outputSpeech": Object {
      "ssml": "<speak> You are not in any active courses.. Anything else? </speak>",
      "type": "SSML",
    },
    "reprompt": Object {
      "outputSpeech": Object {
        "ssml": "<speak> You can ask to list grades, or check if you have any missing assignments. </speak>",
        "type": "SSML",
      },
    },
    "shouldEndSession": false,
  },
  "sessionAttributes": Object {},
  "userAgent": "ask-nodejs/1.0.25 Node/v8.11.4",
  "version": "1.0",
}
`;

exports[`tells you no grades posted yet for a specific class 1`] = `
SkillResponse {
  "response": Object {
    "outputSpeech": Object {
      "ssml": "<speak> For Calculus your grade is not posted yet .. Anything else? </speak>",
      "type": "SSML",
    },
    "reprompt": Object {
      "outputSpeech": Object {
        "ssml": "<speak> You can ask to list grades, or check if you have any missing assignments. </speak>",
        "type": "SSML",
      },
    },
    "shouldEndSession": false,
  },
  "sessionAttributes": Object {},
  "userAgent": "ask-nodejs/1.0.25 Node/v8.11.4",
  "version": "1.0",
}
`;

exports[`tells you your grades for a specific class using computed grades 1`] = `
SkillResponse {
  "response": Object {
    "outputSpeech": Object {
      "ssml": "<speak> For Calculus your grade is A- .. Anything else? </speak>",
      "type": "SSML",
    },
    "reprompt": Object {
      "outputSpeech": Object {
        "ssml": "<speak> You can ask to list grades, or check if you have any missing assignments. </speak>",
        "type": "SSML",
      },
    },
    "shouldEndSession": false,
  },
  "sessionAttributes": Object {},
  "userAgent": "ask-nodejs/1.0.25 Node/v8.11.4",
  "version": "1.0",
}
`;

exports[`tells you your grades for a specific class using computed grades 2`] = `
SkillResponse {
  "response": Object {
    "outputSpeech": Object {
      "ssml": "<speak> For Writing your grade is 91% .. Anything else? </speak>",
      "type": "SSML",
    },
    "reprompt": Object {
      "outputSpeech": Object {
        "ssml": "<speak> You can ask to list grades, or check if you have any missing assignments. </speak>",
        "type": "SSML",
      },
    },
    "shouldEndSession": false,
  },
  "sessionAttributes": Object {},
  "userAgent": "ask-nodejs/1.0.25 Node/v8.11.4",
  "version": "1.0",
}
`;
